src/ws.nim

```diff
@@ -0,0 +1,570 @@
+  result.mask = (b1 and 0x80) == 0x80
+
+  if ws.masked == result.mask:
+    # Server sends unmasked but accepts only masked.
+    # Client sends masked but accepts only unmasked.
+    raise newException(WebSocketError, "Socket mask mismatch")
+
+  var maskKey: seq[byte]
+  if result.mask:
+    # Read the mask.
+    maskKey = await ws.tcpSocket.read(4)
+    if maskKey.len != 4:
+      raise newException(WebSocketError, "Socket closed")
+
+  # Read the data.
+  var data: seq[byte]
+  data = await ws.tcpSocket.read(int finalLen)
+  result.data = toString(data)
+  if result.data.len != int finalLen:
+    raise newException(WebSocketError, "Socket closed")
+
+  if result.mask:
+    # Apply mask, if we need too.
+    for i in 0 ..< result.data.len:
+      result.data[i] = (result.data[i].uint8 xor maskKey[i mod 4].uint8).char
+
+
+proc receivePacket*(ws: WebSocket): Future[(Opcode, string)] {.async.} =
+  ## Wait for a string or binary packet to come in.
+  var frame = await ws.receiveFrame()
+  result[0] = frame.opcode
+  result[1] = frame.data
+  # If there are more parts read and wait for them
+  while frame.fin != true:
+    frame = await ws.receiveFrame()
+    if frame.opcode != Cont:
+      raise newException(WebSocketError, "Socket closed")
+    result[1].add frame.data
+  return
+
+proc receiveStrPacket*(ws: WebSocket): Future[string] {.async.} =
+  ## Wait only for a string packet to come. Errors out on Binary packets.
+  let (opcode, data) = await ws.receivePacket()
+  case opcode:
+    of Text:
+      return data
+    of Binary:
+      raise newException(WebSocketError,
+        "Expected string packet, received binary packet")
+    of Ping:
+      await ws.send(data, Pong)
+    of Pong:
+      discard
+    of Cont:
+      discard
+    of Close:
+      ws.readyState = Closed
+      raise newException(WebSocketError, "Socket closed")
+
+proc sendHTTPResponse*(transp: StreamTransport, version: HttpVersion, code: HttpCode,
+                data: string = ""): Future[bool] {.async.} =
+  var answer = $version
+  answer.add(" ")
+  answer.add($code)
+  answer.add("\r\n")
+  answer.add("Date: " & httpDate() & "\r\n")
+  if len(data) > 0:
+    answer.add("Content-Type: application/json\r\n")
+  answer.add("Content-Length: " & $len(data) & "\r\n")
+  answer.add("\r\n")
+  if len(data) > 0:
+    answer.add(data)
+  try:
+    let res = await transp.write(answer)
+    if res != len(answer):
+      result = false
+    result = true
+  except:
+    result = false
+
+proc validateRequest(transp: StreamTransport,
+                     header: HttpRequestHeader): Future[ReqStatus] {.async.} =
+  if header.meth notin {MethodGet}:
+    # Request method is either PUT or DELETE.
+    debug "GET method is only allowed", address = transp.remoteAddress()
+    if await transp.sendHTTPResponse(header.version, Http405):
+      result = Error
+    else:
+      result = ErrorFailure
+    return
+
+  if header.contentLength() > MaxHttpRequestSize:
```

Comment:
`header.contentLength()` could return `0` and `-1` and pass all the checks here.
https://github.com/status-im/nim-http-utils/blob/master/httputils.nim#L632-L635